doc/adr/0013-provide-a-plugin-api-backed-by-capi-v3

```diff
@@ -0,0 +1,56 @@
+# 13. Provide a Plugin API backed by CAPI V3
+
+Date: 2022-09-16
+
+## Status
+
+Submitted
+
+## Context
+
+_The issue motivating this decision, and any context that influences or constrains the decision._
+
+Although all V8 CLI commands now call [V3](https://v3-apidocs.cloudfoundry.org/) of the CAPI API, CLI plugins are only able to call a Plugin API Interface whose methods invoke a 'legacy' version of the CLI that currently exists in the [`cf`](../../cf/) directory. This is, in effect, V6 of the CLI, which exclusively uses CAPI [V2](https://apidocs.cloudfoundry.org/). The last V6 release was in October 2020, and the `cf` directory has had few commits since then.
+
+The Plugin API exposes ten common CLI commands (e.g. `cf services`) via [methods](https://github.com/cloudfoundry/cli/blob/master/plugin/plugin.go#L17-L48) which return Go structs in the shape of CAPI V2 API resources. Other V6 CLI commands can be invoked by making a [CliCommand](https://github.com/cloudfoundry/cli/blob/d3bb1ecfc94ee12432f8defb30ab9459e83cc727/plugin/cli_connection.go#L72-L74) or [CliCommandWithoutTerminalOutput](https://github.com/cloudfoundry/cli/blob/d3bb1ecfc94ee12432f8defb30ab9459e83cc727/plugin/cli_connection.go#L68-L70), which both take string arguments and return a slice of strings. These methods can be used with `cf curl`, which allows users to make authenticated CAPI API calls, and some plugin authors do so for both [V2](https://github.com/orange-cloudfoundry/cf-plugin-bg-restage/blob/bcb1c4c712a43119d63d849b4b368eb65a86eb3b/repo.go#L72-L80) and [V3](https://github.com/generalmotors/cf-restage-all/blob/d6d9747182881a2b50c7bc601f4762778bb1cbfd/restage_all.go#L211-L237) calls.
+
+In early 2020 the CLI team developed a working POC for a new Plugin API backed by CAPI V3. Although included in beta releases of V7 of the CLI, these changes were mostly [reverted](https://www.pivotaltracker.com/n/projects/892938/stories/171032219) before general availability due to concerns at the scale of breaking changes for plugins, many of which are no longer maintained.
+
+## Decision
+
+_The change that we're proposing or have agreed to implement._
+
+Create a Plugin API V2 that exposes methods that plugins can use to call the same CLI commands used by V8 of the CLI today. Defer making a decision on how long to continue supporting the legacy Plugin API until the uptake of Plugin API V2 is apparent.
+
+The Plugin API V2 will be similar to that developed in early 2020, making use of the same `V7Actor` that V8 of the CLI uses to make V3 CAPI calls. Plugin metadata will declare whether the plugin in question supports the Plugin API V2. If it does not it will be run with the legacy Plugin API.
```

Comment:
I’m unfamiliar with the plugin API that was developed in early 2020 - can y’all provide it for additional context / update this ADR with information surrounding said API?

Without that additional information, I’m left assuming that the proposal is suggesting that the general API remains the same as the current / V1 plugin API, but backed by the V3 API and returns V3 API like objects. This likely means the plugin API will have additional function calls for the newly added V3 resources.

If these assumptions are in line with this proposal than here are my suggestions (if not ignore and I’ll provide an updated response once I have the appropriate information):
1. Switch from existing RPC to a standardized gRPC framework. I honestly don’t remember why this was, I just remember having dozens of conversations with people and the conclusion ending here.
2. Be able to execute [g]RPC calls in a stream-like fashion (ie able to stream STDIN, STDOUT, STDERR as the API server executes requested API calls). This is especially important for plugins that are executing long running functions like `push`.
3. Have a `json` or “structured” output for every CLI function would significantly help plugin authors when parsing the output of certain plugin API calls.
4. _Do not directly tie the V2 Plugin API interface to the V3 API backed actors._ As long as there’s a defined, but separate, plugin API interface the V3 API backed actors can move freely / change without affects / breaking the V2 Plugin API.
5. Indirectly related - when plugins are installed, the `cf` CLI did not keep track of all the necessary information needed for safe / automated upgrade processes. Updating the `cf` CLI to store the following would help with that:
  - Where the plugin was installed from (repo source name, id, etc. Something verifiable?)
  - How the plugin was installed (installed from repo or local or something else?)

Reply:
Thanks for pondering this. For background on the abortive effort in early 2020, see [this epic](https://www.pivotaltracker.com/n/projects/892938/epics/4315861). The last commit before those changes were reverted was [this one](https://github.com/cloudfoundry/cli/tree/0d0655662c3143a6275dd7feed1ca40a0d2f89a2). [This doc](https://github.com/cloudfoundry/cli/blob/master/doc/v7-plugin-api-changes.md) also described some of the changes in the new Plugin API.

Having done worked only a tiny bit with gRPC before, I'll defer to others on this, but I do like the idea of being able to generate server and client-side code. And that could help achieve 3. - I agree that it'd be a really good thing to totally eliminate the need to parse responses as slices of strings.